core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java

```diff
@@ -0,0 +1,1331 @@
+            long tmpHigh = negateHighExact(high, low);
+
+            low = tmpLow;
+            high = tmpHigh;
+        }
+
+        // Scales down for 10**rescaleFactor.
+        // Because divide by int has limited divisor, we choose code path with the least amount of divisions
+        if ((scaleFactor - 1) / MAX_POWER_OF_FIVE_INT < (scaleFactor - 1) / MAX_POWER_OF_TEN_INT) {
+            // scale down for 10**rescale is equivalent to scaling down with 5**rescaleFactor first, then with 2**rescaleFactor
+            scaleDownFive(high, low, scaleFactor, result, offset);
+            shiftRight(result[offset], result[offset + 1], scaleFactor, roundUp, result, offset);
+        }
+        else {
+            scaleDownTen(high, low, scaleFactor, result, offset, roundUp);
+        }
+
+        if (negative) {
+            // negateExact not needed since all positive values can be negated without overflow
+            negate(result, offset);
+        }
+    }
+
+    /**
+     * Scale down the value for 5**fiveScale (result := decimal / 5**fiveScale).
+     */
+    private static void scaleDownFive(long high, long low, int fiveScale, long[] result, int offset)
+    {
+        while (true) {
+            int powerFive = Math.min(fiveScale, MAX_POWER_OF_FIVE_INT);
+            fiveScale -= powerFive;
+
+            int divisor = POWERS_OF_FIVES_INT[powerFive];
+            dividePositives(high, low, divisor, result, offset);
+
+            if (fiveScale == 0) {
+                return;
+            }
+
+            high = result[offset];
+            low = result[offset + 1];
+        }
+    }
+
+    /**
+     * Scale down the value for 10**tenScale (this := this / 5**tenScale). This
+     * method rounds-up, eg 44/10=4, 44/10=5.
+     */
+    private static void scaleDownTen(long high, long low, int tenScale, long[] result, int offset, boolean roundUp)
+    {
+        if (high < 0) {
+            throw new IllegalArgumentException("Value must be positive");
+        }
+
+        boolean needsRounding;
+        do {
+            int powerTen = Math.min(tenScale, MAX_POWER_OF_TEN_INT);
+            tenScale -= powerTen;
+
+            int divisor = POWERS_OF_TEN_INT[powerTen];
+            needsRounding = divideCheckRound(high, low, divisor, result, offset);
+
+            high = result[offset];
+            low = result[offset + 1];
+        }
+        while (tenScale > 0);
+
+        if (roundUp && needsRounding) {
+            incrementUnsafe(result);
+        }
+    }
+
+    static void shiftRight(long high, long low, int shift, boolean roundUp, long[] result, int offset)
+    {
+        if (high < 0) {
+            throw new IllegalArgumentException("Value must be positive");
+        }
+
+        if (shift == 0) {
+            return;
+        }
+
+        boolean needsRounding;
+        if (shift < 64) {
+            needsRounding = roundUp && (low & 1L << (shift - 1)) != 0;
+
+            low = (high << 1 << ~shift) | (low >>> shift);
```

Comment:
Why double `<<` instead of `<< (1 - shift)`?

Reply:
Actually, I think this is just equivalent to `-shift`:

`-shift = ~shift + 1` => `~shift = -shift - 1` => `(high << 1 << ~shift) == high << 1 - shift - 1 == high << -shift`.

However, I need to check if it's equivalent for some of the edge cases of shifts in java, which is really: `v >> s == v >> (s % 64)` (e.g., `-1 >> 64 == -1`, `16L >> 65 == 8`, etc.)

Reply:
Indeed,

```
1. high << 1 << ~shift
```

and

```
2. high << 1 + (-shift - 1) 
    == high << 1 - shift - 1
    == high << -shift 
    == high << 64 - shift
```

are not quite equivalent.

When `shift == 0`

```
1. high << 1 << -1 == high << 1 << 64 - 1 == high << 1 << 63
```

But,

```
2. high << 64 - 0 == high << 64 == high << 0
```